libs/depgraph/depgraph_api.go

```diff
@@ -0,0 +1,608 @@
+// Copyright (c) 2022 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package depgraph
+
+// Graph is a dependency graph.
+// The main use-case is to represent configuration items (network interfaces, routes,
+// volumes, etc.) or any managed stateful objects (incl. processes, containers, files,
+// etc.) as graph nodes and their dependencies as directed graph edges.
+// For more information please see README.md.
+type Graph interface {
+	GraphR
+
+	// SetDescription updates description assigned to the (sub)graph.
+	SetDescription(string)
+
+	// PutNode adds or moves (and updates) node into this (sub)graph.
+	PutNode(node *Node)
+	// DelNode deletes an existing node from this (sub)graph.
+	// Returns true if the node existed and was actually deleted.
+	DelNode(NodeID) bool
+
+	// PutSubGraph adds a new subgraph into this graph or updates an existing
+	// subgraph. This refers to a direct child of this graph, cannot add/update
+	// a nested subgraphs.
+	PutSubGraph(Graph)
+	// DelSubGraph deletes existing subgraph. This refers to a direct child of this
+	// graph, cannot delete a nested subgraph.
+	// Returns true if the subgraph existed and was actually deleted.
+	// It is an error to try to use a subgraph after it was deleted (can't be used
+	// even as a separate graph anymore).
+	DelSubGraph(name string) bool
+	// EditSubGraph elevates read-only subgraph handle to read-write access.
+	// Panics if the given graph is not actually a subgraph (direct or nested)
+	// of this graph.
+	EditSubGraph(GraphR) Graph
+	// EditParentGraph returns read-write handle to a (direct) parent graph
+	// of this subgraph.
+	// Return nil if the graph is not a subgraph.
+	EditParentGraph() Graph
+
+	// PutPrivateData allows the user to store any data with the graph.
+	// Note that it is also possible to store private data next to each node.
+	// (see Node.PrivateData)
+	PutPrivateData(interface{})
```

Comment:
What is this for?

Reply:
The reconciler needs to have some context alongside the current-state graph to manage async operations. Also it needs to have the ability to resume reconciliation after an async operation(s) finalized. My goal here (and with the Reconciler API) was to have the reconciler stateless, so that you can throw it out even if some async operations are still running. Also the reconciliation is run in the same Go routine as the one from which it was called (apart from async operations). So no Go routines are spawned by Reconciler (only by Configurators maybe for async ops). So when `Reconcile()` returns, you know that no reconciliation is ongoing, no one else is reading/writing the current/intended state graphs, and you can start next reconciliation (with possibly different intended state and a new instance of reconciler) whenever you want. Using this context stored alongside the current-state graph, the reconciler will be aware of async operations from previous reconciliations and their current status. I found this as less bug-prone even if it maybe diverges from common promise-style engines.

Anyway, say we switch to your suggested API:

> `ch := r.Reconcile(ctx, currentState, intendedState)`

To do this, the reconciler would start the reconciliation inside a Go routine and return immediately.
The Go routine could stream status of executed operations through the channel.

The challenges here are:
- What if I want to update the intendedState (I received an update from controller)? Can I update `intendedState` even if there is a Reconciler Go routine still possibly running and reading the graph?
- Can I trigger another reconciliation while the previous one is still running? If yes, then the next Reconcile needs to somehow communicate with the Go routine of the previous one and cooperate to avoid race conditions.

Reply:
> So, private data is a current state from a reconciler

Yes

>  I can sort of do:
>
>   1. start a reconciliation
>   2. export the state
>   3. stop a reconciliation
>   4. start a new reconciliation importing the state
>   5. keep going?

With the current API this is simplified for the user to just:
1. run a reconciliation
2. do something else, maybe update the intended state
3. run a reconciliation
4. keep going

I didn't want to have the reconciliation context separate from the graph to decrease the chance of bugs (like lost context, or used old obsolete context, etc.).
Also the user doesn't have to keep track of previous reconciliations, he does not have to stop them, graphs do not have to be locked, etc.
I think that with the current API it will be simpler, but at this point the discussion is only theoretical. I would like to first use the reconciler in EVE, get some experience and maybe then change the API if it is not very good to work with. Here we are talking about the return value of only a single function (`Reconcile()`), so making modifications to it in the future should not be so drastic.

Reply:
I still don't quite get it, but if it is something we don't even know that we will use, then sure, start with it simple, we can gain experience.